### PR TITLE
fix: add ability to filter transfers by customerIDs query param

### LINKED ISSUE #582: CustomerIDs query param not being honored
According to the API docs (https://api.moov.io/v1/#operation/getTransfers) the `GET /transfers` endpoint should accept an optional query param for `customerIDs` which would limit the results of the request to any transfer where that customerID was in the source or destination.

This is not functioning as expected today. The result is not filtered.

### PR DESCRIPTION
**Pull request by vxio**
Fixes #582

### CODE REVIEW

File: pkg/transfers/router.go [hunk 1]
```diff
@@ -123,6 +126,9 @@ func readTransferFilterParams(r *http.Request) transferFilterParams {
 		if s := strings.TrimSpace(q.Get("status")); s != "" {
 			params.Status = client.TransferStatus(s)
 		}
+		if ids := q.Get("customerIDs"); ids != "" {
+			params.CustomerIDs = strings.Split(ids, ",")
```
Comment (adamdecaf) [inline pkg/transfers/router.go:130]: Should we put a limit on the number of customerIDs allowed? 25? 50?

**Review by adamdecaf — APPROVED** [review]

**Comment by adamdecaf** [comment]
@joshsadler any thoughts on what a maximum number of customerIDs allowed in `GET /transfers?customerIDs=...` could be?

File: pkg/transfers/repository_test.go [hunk 2]
```diff
@@ -58,6 +59,46 @@ func TestRepository__getTransfersWithTraceNumbers(t *testing.T) {
...
+	got, err := repo.getTransfers(namespace, params)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(got) != len(params.CustomerIDs) {
+		t.Fatalf("# of transfers: want %d, got %d", len(params.CustomerIDs), len(got))
+	}
+
+	want := make(map[string]bool, 0)
```
Comment (adamdecaf) [inline pkg/transfers/repository_test.go:88]: From CI: 

> pkg/transfers/repository_test.go:88:32: should use make(map[string]bool) instead (S1019)

**Comment by joshsadler** [comment]
> @joshsadler any thoughts on what a maximum number of customerIDs allowed in `GET /transfers?customerIDs=...` could be?

@adamdecaf I don't have a strong use case for multiple customerIDs in the product today. If you need an answer right away I'd go with ~25 which safely keeps the URL below the string size limit of old browsers.

**Comment by vxio** [comment]
@adamdecaf @joshsadler Ok I'll set `25` for now, we can always tweak it in the future.

File: pkg/transfers/repository.go [hunk 3]
```diff
@@ -48,19 +49,39 @@ func (r *sqlRepo) getTransfers(namespace string, params transferFilterParams) ([
...
 		return nil, err
 	}
 	defer stmt.Close()
 
-	args := []interface{}{namespace, params.StartDate, params.EndDate, params.Count, params.Skip}
+	args := []interface{}{namespace, params.StartDate, params.EndDate}
 	if statusQuery != "" {
 		args = append(args, params.Status)
 	}
+	if len(params.CustomerIDs) > 0 {
```
[the remark this answers is not in the document]
Comment (vxio) [reply pkg/transfers/repository.go:76]: No, unfortunately, you'll get this compile error: 
image

File: pkg/transfers/repository.go [hunk 4]
```diff
@@ -48,19 +49,39 @@ func (r *sqlRepo) getTransfers(namespace string, params transferFilterParams) ([
 	if string(params.Status) != "" {
 		statusQuery = "and status = ?"
 	}
-	query := fmt.Sprintf(`select transfer_id from transfers
-where namespace = ? and created_at >= ? and created_at <= ? and deleted_at is null %s
-order by created_at desc limit ? offset ?;`, statusQuery)
+	var customerIDsQuery string
+	if len(params.CustomerIDs) > 0 {
```
[the remark this answers is not in the document]
Comment (vxio) [reply pkg/transfers/repository.go:60]: I like the idea! Check out my updated code. The downside is the query string is somewhat spread throughout the block, but it's not too bad.

File: pkg/client/api/openapi.yaml [hunk 5]
```diff
@@ -218,6 +218,7 @@ paths:
         required: false
         schema:
           example: c336f57e,476547a8
+          maxItems: 25
```
Comment (vxio) [inline pkg/client/api/openapi.yaml:221]: Is `maxItems` the right key here?  
The query param is a string of a list of ids: `?customerIDs=id1,id2,...`

Comment (vxio) [reply pkg/client/api/openapi.yaml:221]: Actually, I should tweak the schema here:
```
schema:
  type: array
  items:
    type: string
...
```
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml#L89

Comment (adamdecaf) [reply pkg/client/api/openapi.yaml:221]: There's a difference between these three. We want the first.

- `?customerIDs=c1,c2,c3` (`type: string`)
- `?customerIDs=c1&customerIDs=c2` (`type: array`)
- `?customerIDs=[c1,c2,c3]` 
- `?customerIDs[]=c1,c2,c3`

Comment (vxio) [reply pkg/client/api/openapi.yaml:221]: Oof, my bad, thanks for the clarification!

And would the `maxItems: 25` be ok for the first one? I can't find anything in the swagger spec to confirm this.

Comment (adamdecaf) [reply pkg/client/api/openapi.yaml:221]: I think `maxItems: <number>` is only valid on `type: array`, but this parameter is `type: string`.

Comment (vxio) [reply pkg/client/api/openapi.yaml:221]: Hmm `make client` worked; do you recommend just specifying this limit in the description?

Comment (adamdecaf) [reply pkg/client/api/openapi.yaml:221]: Yea. We probably want to reject requests with 26 customerID's, but that seems annoying for callers.

Comment (vxio) [reply pkg/client/api/openapi.yaml]: Updated the docs here.

**Review by adamdecaf — APPROVED** [review]

**Pull request merged**